zero/src/prover.rs

```diff
@@ -131,29 +135,180 @@ impl BlockProverInput {
+                            break;
+                        }
+                        let seg_prove_ops = seg_prove_ops.clone();
+                        let proof_runtime = proof_runtime.clone();
+                        let segment_proof_tx = segment_proof_tx.clone();
+                        // Prove one segment in a dedicated async task.
+                        let segment_proving_task = tokio::spawn(async move {
+                            if let Some(segment_data) = segment_data {
+                                debug!("proving the batch {batch_idx} segment nr. {segment_counter}");
+                                let seg_aggregatable_proof= Directive::map(
+                                    IndexedStream::from([segment_data]),
+                                    &seg_prove_ops,
+                                )
+                                    .run(&proof_runtime.heavy_proof)
+                                    .await?
+                                    .into_values_sorted()
+                                    .await?
+                                    .into_iter()
+                                    .next()
+                                    .context(format!(
+                                        "failed to get segment proof, batch: {batch_idx}, segment: {segment_counter}"
+                                    ))?;
+
+                                segment_proof_tx
+                                    .send((segment_counter, seg_aggregatable_proof))
+                                    .await
+                                    .context(format!(
+                                        "unable to send segment proof, batch: {batch_idx}, segment: {segment_counter}"
+                                    ))?;
+                            };
+                            Ok::<(), anyhow::Error>(())
+                        });
+
+                        segment_proving_tasks.push(segment_proving_task);
+                        segment_counter += 1;
+                    }
+                    drop(segment_proof_tx);
+                    // Wait for all the segment proving tasks of one batch to finish.
+                    while let Some((segment_idx, segment_aggregatable_proof)) = segment_proof_rx.recv().await {
+                        batch_segment_aggregatable_proofs.push((segment_idx, segment_aggregatable_proof));
+                    }
+                    try_join_all(segment_proving_tasks).await?;
+                    batch_segment_aggregatable_proofs.sort_by(|(a, _), (b, _)| a.cmp(b));
+                    debug!(block_number=%block_number, batch=%batch_idx, "finished proving all segments");
+                    // We have proved all the segments in a batch,
+                    // now we need to aggregate them to the batch proof.
+                    // Fold the segment aggregated proof stream into a single batch proof.
+                    let batch_proof = if batch_segment_aggregatable_proofs.len() == 1 {
```

Comment:
nit: usecase for [`itertools::Itertools::exactly_one`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.exactly_one).
Handle the error case with your `Directive::fold` code, move the `map(|(_, it)| it)` higher up, and get rid of the `unwrap()`

Reply:
This is a valid suggestion. By other view, `exactly_one` semantic is a bit misleading to be used here. `exactly_one`  implies that iterator is expected to be size one, and if not that is the error in the flow that we need to handle (and that is not the case here).

Reply:
Btw. I can not move the mapping before ` Directive::fold`, as it only works with `SegmentAggregatableProof` sa inputs.